Source/Fuse.Scripting.JavaScript/JavaScript.Model.uno

```diff
@@ -0,0 +1,254 @@
+        [UXAttachedPropertySetter("Model"), UXNameScope]
+        public static void SetAppModel(IApp app, IExpression model)
+        {
+			app.Root.RemoveAllChildren<ModelJavaScript>();
+
+			var _appModel = ModelJavaScript.CreateFromPreviewState(app.Root, model);
+			app.Root.Children.Add(_appModel);
+        }
+	}
+	
+	class ModelJavaScript : JavaScript, IPreviewStateSaver
+	{
+		internal class ParsedModelExpression
+		{
+			public IExpression Source;
+			public string ModuleName;
+			public string ClassName;
+			public List<string> Args = new List<string>();
+			public List<Dependency> Dependencies = new List<Dependency>();
+			
+			public string ArgString
+			{
+				get
+				{
+					var str = "";
+					for (int i=0; i < Args.Count; ++i)
+						str += ", " + Args[i];
+					return str;
+				}
+			}
+			
+			public bool CompatibleTo( ParsedModelExpression o )
+			{
+				//there is no way to migrate these now as they might refer to tree objects, thus reject entirely
+				if (Args.Count != 0 || o.Args.Count != 0 ||	
+					Dependencies.Count != 0 || o.Dependencies.Count != 0)
+					return false;
+					
+				return o.ModuleName == ModuleName &&
+					o.ClassName == ClassName;
+			}
+		}
+		
+		internal static ParsedModelExpression ParseModelExpression(IExpression exp, NameTable nt)
+        {
+            if (exp is Data) {
+                var className = ((Data)exp).Key;
+                return new ParsedModelExpression{ ClassName = className, ModuleName = className,
+					Source = exp };
+            } 
+            else if (exp is Divide)
+            {
+                var left = ParseModelExpression(((Divide)exp).Left, nt);
+                var right = ParseModelExpression(((Divide)exp).Right, nt);
+                
+                if (left.Args.Count > 0 || left.Dependencies.Count > 0)
+					throw new Exception( "Invalid Model path expression: " + exp);
+                
+                right.ModuleName = left.ModuleName + "/" + right.ModuleName;
+                right.Source = exp;
+                return right;
+            }
+            else if (exp is Fuse.Reactive.NamedFunctionCall)
+            {
+                var nfc = (Fuse.Reactive.NamedFunctionCall)exp;
+
+                var result = new ParsedModelExpression{ ClassName = nfc.Name, ModuleName = nfc.Name,
+					Source = exp };
+                for (int i = 0; i < nfc.Arguments.Count; i++)
+                {
+					var argName = "__dep" + i;
+					var c = nfc.Arguments[i] as Constant;
+                    result.Dependencies.Add(new Dependency(argName, nfc.Arguments[i]));
+					result.Args.Add( argName );
+                }
+
+                return result;
+            }
+            else throw new Exception("Invalid Model path expression: " + exp);
+        }
+
+        void SetupModel()
+        {
+			if (_model == null)
+			{
+				Code = "";
+				return;
+			}
+
+            //TODO: this should not be necessary. It's done because there's a UX processor error, we get
```

Comment:
The problem here goes beyond the TODO. It shouldn't be getting an expression at all, it's rebuilding a scoped name from an expression (a hack). If that were fixed then the TODO wouldn't be an issue -- nothing should be looking at the structure of an IExpression!

Reply:
> nothing should be looking at the structure of an IExpression!

Absolutely, we need to be able to change how a given expression is compiled/evaluated, even if when evaluated it will provide the same result (which is of course a bit orthogonal to this use case).